cpp_examples/rhi_examples/CMakeLists.txt

```diff
@@ -1,6 +1,6 @@
 macro(make_sample executable_name src_file)
 add_executable(${executable_name})
-set_property(TARGET ${executable_name} PROPERTY CXX_STANDARD 17)
+set_property(TARGET ${executable_name} PROPERTY CXX_STANDARD 20)
```

Comment:
So this is an example, there's no "user" for this code. As for which feature it's using, this code is using designated initializers

Reply:
On a second thought, I still feel it's better to keep consistent  CXX_STANDARD across the entire project. Will it be a big blocker to you if we stick to c++17 for now?

Reply:
I think as example code the C++ 20 syntax is easier to read and learn. So for that purpose I wanted to use C++ 20

Reply:
Do agree that `designated initializers` makes the code look better and it is a nice-to-have feature. 

However, C++20 features as a whole haven't been well supported on main stream compilers for now. As you may notice from https://en.cppreference.com/w/cpp/compiler_support, there are recognizable diffs between different compilers/compiler-versions. I'm a bit worried if it gonna cause `rhi_examples` harder to maintain especially in terms of cross-platform supports.

@PENGUINLIONG  @ailzhang  WDYT?

Reply:
A compiler does not need to implement the full C++ 20 specs to turn on the support. In this case designated initializers are fully supported from GCC 8 / Clang 10 / MSVC 19.21 which are available or required on all of our platforms already... Only missing thing is macOS 10.14, where I don't think we have MoltenVK backend enabled at all

Reply:
`designated initializers` isn't the concern. The problem is once we decide to turn on C++20 in the project, people may start using other C++20 features in `rhi_examples` including the unstable ones. IMHO, it's better to plan C++ upgrade across the entire project so fundamental dependencies (libc++.so, etc) can be consistent across different executable/shared-libraries targets.

However, I understand this is controversial and even myself is a bit swing in the case of `cpp_examples`, so sincerely asking advice from @PENGUINLIONG @ailzhang

Reply:
Hi thanks for the thorough discussion here! Adding my $0.02:
- We've already deprecated support on macos10.14 so that's no longer an issue. 
- But IMHO we should use C++17 for now and add a TODO saying this can be much simpler using `designated initializers` after we upgrade to C++20. I agree with @jim19930609 that by setting C++20 here we're opening a potential risk for developers to use other C++20 features, which may not be as widely supported in mainstream compilers as `designated initializers` and it's better practice to discuss & upgrade C++ standard project wide separately. 
- As of `rhi_example` target itself, TBH it's not built by default so it's kinda fragile at this moment. These examples may not be user facing atm. But to make it a helpful reference for devs/users in the future we'd still like it to comply with the project-wide setting to minimize maintenance cost. 
- Currently Taichi is pretty flexible in what host compiler to use and we didn't really force a minimal clang/gcc/msvc version other than C++17 standard. 

If the rhi_examples is really non-user facing and non-dev facing in the future, we should discuss how to maintain it (e.g. how to make sure it's working, whether keep it in taichi repo or a separate one, use C++17 or C++20 etc), and that's definitely out of scope for this PR which tries to enable MSBuild for taichi. @bobcao3 @jim19930609 wdyt?

Reply:
So if I'm understanding this correctly, we will be fixing the C++20 syntax in this PR? (It seems that Clang / GCC builds the examples without complaining but MSVC is a lot stricter (considering we have strong warning flags)

Reply:
Yup that'd be great! I guess we can change the syntax in this PR to make MSVC happy with C++17 at least (or if there're too many we may allow a few temporarily and fix them in followup PRs.

Reply:
```suggestion
set_property(TARGET ${executable_name} PROPERTY CXX_STANDARD 17)
```